docs: Ahead of time Libvirt defaulting on the cluster/API level

rmohr opened the pull request:
This Proposal fills a gap between KubeVirt as Runtime and Higher level management, like proposed in #272. Without the necessity for asynchronous specification updates, we come much closer to the meaning of a Pod specification in Kubernetes, but most important, this allows decoupling administrative tasks completely from the runtime (the need to run VMs, to get a full VM spec), when manipulating VM Specs.

stu-gott commented on docs/defaulting.md:
```diff
@@ -0,0 +1,128 @@
+
+In order to make sure that the Libvirt version, which does the defaulting
+generates appliable defaults, it must first be compatible with the libvirt/qemu
+pairs, currently running on the hostr. Second a guest machine type must be
```
"host" vs "hostr"

stu-gott commented on docs/defaulting.md:
```diff
@@ -0,0 +1,128 @@
+
+### Defaulting service based on libvirt
+
+Except for some runtime information, which is normally not included in migratable domain XMLs, libvirt fills in all defaults when defining a domain. We can leverage that with a simple REST based service ,which defines VMs in an isolated libvirt (running libvirt and the small REST service in a Pod). The service is a standalone component in KubeVirt and can then be integrated into two different flo ...
```
misplaced comma after service. "service, " vs "service ,"

davidvossel commented:
In PR 287, we make a distinction between expansion and defaulting. https://github.com/kubevirt/kubevirt/pull/287

In this PR it appears expansion and defaulting are just referred to as defaulting. I'm a little confused by the terminology.

davidvossel commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+isolated libvirt (running libvirt and the small REST service in a Pod). The
+service is a standalone component in KubeVirt and can then be integrated into
+two different flows:
+
```
We have two objects that we've agreed on 

VM - represents spec of a running instance
VMConfig - represents spec of a config that can be used to start an instance.

So, /start on a VMConfig creates a VM

Just to be clear, when we are talking about introducing this libvirt defaulting service, what is the service acting on?  Is it filling in the spec for the VMConfig or VM? The fact I'm asking this probably means we need more explicit terminology.

rmohr replied:
This only deals with VMs. The runtime does not know anything about VMConfigs. The runtime does also not know how to convert from a VM to a VMConfig, or the other way round. The implementation of /start in the VMConfig store will have to deal with that conversion.

michalskrivanek replied:
then why is it needed? I still find the concept of a "libvirt dry run" to fill a host-specific default weird. Especially since not all the things are filled by libvirt (e.g. the example with qxl and spice - that's not done by libvirt)

berrange replied:
You say that the runtime doesn't know  how to convert between VM & VMConfig. I don't think that should even be a problem that needs solving - the DomainSpec sub-object schema to describe configuration should be common between both of these types.

rmohr replied:
Right. I agree. I just wanted to highlight the point.

rmohr replied:
@michalskrivanek libvirt adds a video device, but you are right it is not qxl, it adds cirrus by default. Will update that.

michalskrivanek replied:
I think we need a more comprehensive solution - as I hinted in another PR - something similar to the logic of virt-install/virt-xml to simplify the VMConf->VM process. But I wonder if we really have to strive for the complete VM spec ahead of time - those details should not be needed by the upper layer - it should deal with it and treat it transparently (preserve PCI addresses if you don't change the VMConf, but wipe it out and let libvirt construct it again when you e.g. remove & add a device triggering also a creation of an additional bridge.
Libvirt is not going to help you with most of that logic

rmohr replied:
@michalskrivanek  I don't think that this proposal here  is in the way of having more sophisticated logic on top (also in the runtime). libosinfo, presets, flavours, whatever ... can all be applied in a chain, before this stage is reached.

Could you further explain the bridge example? It sounds interesting and I don't yet understand, why it is ok in this case, to remove the PCI addresses which were already assigned ...

davidvossel replied:
> This only deals with VMs. The runtime does not know anything about VMConfigs. The runtime does also not know how to convert from a VM to a VMConfig, or the other way round. The implementation of /start in the VMConfig store will have to deal with that conversion.

If this proposal is only working on the VM during creation time, it is unclear to me how this service will help us in the case where we want to start a VM, stop it, then start it again and ensure the same expanded VM spec is used the second time.

rmohr commented:
@davidvossel While there can be other admission controllers (https://kubernetes.io/docs/admin/admission-controllers/) later (e.g. flavours, presets, ...), and they all do some expansion, this one would be the first controller, which expands the spec.

Some thoughts about differences to #287: 

 * This is purely about runtime
 * In the context of the VM POST/PUT call, this defaulting would be the first admission controller we have . In that context it gives you a fully, with defaults populated spec, as the direct response from the creation of the VM in the runtime.
 * The dedicated REST endpoint would allow this or more admission controller to expand the config without the need to run it.

Does that answer why this defaulting talks about expansion?

davidvossel commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+    specification. This updated spec contains all the necessary bits, to allow
+    consistent hotplugging and keeping the VM definition consistent between
+    redefines.
+ 2. Add an extra endpoint to `virt-api` which only purpose is to prefill the VM
```
I want to make sure I understand the two options here.

Option 1. automatically prefills in defaults when a VMConfig is posted for the first time.

Option 2. Provides an endpoint for users to post a VMConfig to and receive the populated VMConfig back. 

If my understanding is correct, with option 1 we attempt to solve the issue within kubevirt automatically, With option 2 we push this logic to the administrator and provide an endpoint to help them with populating the VM spec.

If we can abstract away this complexity from the administrator, that is my preference (option 1).  What is the concern about hotplugging with option 1? Understanding the flow someone would go through to hotplug a device might make this more obvious. It is unclear to me what object (VMConfig or VM) someone is modifying to hotplug.

rmohr replied:
Ok, so option 1 is great, if you don't care about UIs, or anything. It is like with the Kubernetes runtime. You have your VM spec somewhere, and want to start your VM. You post it, and as a result  you get a full spec back. Additionally, sooner or later your VM will run. Also hotplug, ... can just be done by modifying the object.

However, if you want to build something like nice management UIs on top, and you want to support hotplug and modifications for next runs in combination with a consistend VM Spec, it gets more difficult. 
Your management app would have to deal with VM Spec which are not yet fully defaulted, because they never ran. If libvirt on the node adds relevant spec info for persistend VM specs, you have asynchronous spec updates, which you have to consider. If you hotplug a sparse device, which you want to keep for the next run, you need the fully populated device after it was added by libvirt. 

That all can be worked around by the endpoint in option 2. It allows you to always fill out all the relevant data like PCI addresses, without every having to even ru n a VM. I don't see option 2 as something important, if you use the runtime directly without a management app on top, but it could still be useful.

Does that make sense?

davidvossel replied:
Ok I think I understand the crux of the issue here. We're bouncing between two concepts for starting/managing a VM.

1. (the current way) Post a VM definition to the cluster
2. (yet to be implemented way) Post a VirtualMachineConfig to the cluster and call /start and /stop on that config.

Once we go to item 2's workflow, we will not be operating using the same pattern as pods. it's a weird concept to say, start a VM using the VM config, but hot plug devices using the VM object directly.  From the user perspective I think there needs to be a single entry point.

If we're saying people can start/stop a VM using the VM config, I think we also need to say users can hotplug devices by modifying the VMConfig.

So all that said... I *think* this means the defaulting service should work on VMConfigs as they are posted to the cluster.

davidvossel commented:
> @davidvossel While there can be other admission controllers (https://kubernetes.io/docs/admin/admission-controllers/) later (e.g. flavours, presets, ...), and they all do some expansion, this one would be the first controller, which expands the spec.

> Does that answer why this defaulting talks about expansion? I didn't have the time to properly look into #287. It was added after this one.

expansion in #287 is defined as the process of adding required dependencies to the spec. Defaulting is defined as adding a default device for something in the spec.

In this pull request, I'm not seeing consistent language used based on the definitions in #287.  Defaulting and expansion appear to be separate concepts, but this PR appears to be wrapping the "expansion" process into "defaulting".  Our terminology here is confusing. (maybe it's just me?)

rmohr commented:
@davidvossel 

> One issue I'm encountering with this proposal is that I understand the concepts of what is being proposed here, but the mechanics of exactly what objects are being worked on are not obvious to me.

This proposal only knows about VMs, I will try to better describe it, but the two options talk about VMs exclusively. The VirtualMachineConfig in #272 is not part of the runtime (will probably not even be in the same git repo). It is not clear to me what #287 tries to achieve (if it tries to solve things for the runtime or for the external VM store). At the current stage the config is an alternative to storing your VMs for instance in git. The VM config/VM store is more comparable to helm or the service-catalog.

rmohr commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+### Defaulting service based on libvirt
+
+Except for some runtime information, which is normally not included in
+migratable domain XMLs, libvirt fills in all defaults when defining a domain.
```
When using the terms of #287 it would do both. Whatever libvirt uses as it's default values (e.g. you didn't specify a PCI address for your device, it will fill in the address for you) and default devices (e.g. if you only add a spice device, it will automatically add a qxl device for you).

rmohr replied:
I think https://github.com/kubevirt/kubevirt/pull/287#discussion_r126257956 could also be useful where I would see this proposal in the stack.

michalskrivanek replied:
I think we should try not to require PCI addresses (and in general the need to go to libvirt) in VM spec. I agree it should be complete at the definition time (by admission controller - flavor, presets, whatever service) but hopefully we can avoid the host level of details in the spec itself.

I.e. kubevirt VM spec should abstract all the internal system devices which kubevirt users do not care about, then we do not need to have any explicit libvirt defaulting for them, we don't need to back-propagate them into the VM spec.

michalskrivanek replied:
@berrange, yes, I agree that the guest layout is our contract, that's what we need to keep unchanged between restarts, migrations, as much as we can.

I see a need for a service/lib analogous to virt-xml and libosinfo  to be able to "expand", but we shouldn't need to express anything in VM spec for internal devices, things we do not want to expose to kubevirt users. 
So...do you see the need to expose PCI addresses to kubevirt users? Do you think those details should necessarily be in VM spec? If so, why? Aren't we trying to converge to a higher level of abstraction (sure, as long as we keep the guest machine ABI unchanged, but that can be just an implementation detail)

I would expect something more friendly, e.g. for NICs something like
nic1: 
    mac: mac1
nic2:
    mac: mac2

to be able to express order in guest, but why addresses? Again, as long as they are kept the same on restarts.

rmohr commented:
I have now a better understanding about the terminology in #287 and will try to incorporate it here, to make it better understandable where this admission controller would fit in in a set of additional admission controllers.

rmohr commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+   volume claim. In such cases most of the time, the device `source` section
+   needs to be replaced by a dummy `source` inside the defaulter. Then the
+   fully populated `target` section needs to be mapped back.
+ * There exist mandatory Domain XML fields, which we can't allow to be set on
```
That is what the whole paragraph says.

fabiand replied:
Urgh - You are right, not sure how I misread it :)

rmohr commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+ * The Kubevirt VM spec can contain abstract definitions, like a reference to a
+   volume claim. In such cases most of the time, the device `source` section
+   needs to be replaced by a dummy `source` inside the defaulter. Then the
+   fully populated `target` section needs to be mapped back.
```
Will add one

rmohr commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+device has changed, or needs an update, you would have to rely on the order of
+the devices (which is not necessarily stable:
+https://www.redhat.com/archives/libvirt-users/2012-December/msg00087.html), or
+try to match them based on their content.
```
that is something I would want to have. So that, management outside of KubeVirt does not have to care about detecting the right devices.

rmohr commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+generates appliable defaults, it must first be compatible with the libvirt/qemu
+pairs, currently running on the host. Second a guest machine type must be
+specified, to guarantee migratability between the different versions in the
+cluster.
```
First I think it is outside of this proposal, since we don't guarantee that right now. Second, I think, having something like a configmap, which contains for instance a namespace or cluser wide guest machine would do the trick. Like you do it right now in oVirt. You tell your cluster up front what to use for these values. If we want to enforce this or not, depends if we want to allow creating cluster which do or don't care about migrations.

rmohr commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+
+ * Hotplug operations are not possible with the minimal VM spec of kubevirt,
+   since we could match wrong devices.
+ * After VM restarts, different devices or addresses could be assigned to a VM
```
Yes. In this case, you just don't post the fully expanded spec when you want to start the VM again. I think that is covered.

For the defaulting phase, when starting, it is sufficient to have a cluster compatible, but new enough libvirt involved, to get the newest and best default values. that can be easily guaranteed by updating the defaulting service first, when you update kubevirt.

rmohr commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+Libvirt allows you to post very minimal Domain specifications. It will then
+fill out a lot of device details for you. On cluster level this can lead to
+difficulties, since depending on the fact if the VM has ever run before, you
+will have to operate on different specs, and you have to incoroporate VM
```
I will rephrase that. That is indeed a little bit vague.

rmohr commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+populating a VM specification, we come much closer to the meaning of a Pod
+specification in Kubernetes, but most important, this allows decoupling
+administrative tasks completely from the runtime, when manipulating VM Specs
+(the need to run VMs, to get a full VM spec).
```
The title is good. It actually replaces the defaulting of libvirt on the node level.

rmohr commented on docs/defaulting.md:
```diff
@@ -0,0 +1,180 @@
+## Current situation
+
+Libvirt allows you to post very minimal Domain specifications. It will then
+fill out a lot of device details for you. On cluster level this can lead to
```
I don't think expand should have any more meaning than filling  out.

rmohr commented:
Added implementation details, and explain how this admission-controller would work in a stack of admission controllers.

rmohr commented:
Removed the more advanced use case, and added additional descriptions. It should now be more focused and easier to understand why it is there. Please have another look.

davidvossel commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+   volume claim, or a reference to a host network. In such cases most of the
+   time, the device `source` section needs to be replaced by a dummy `source`
+   inside the defaulter. Then the fully populated `target` section needs to be
+   mapped back (e.g. a PVC does not have to exist, when posting a VM. It can be
```
so, just to be clear.   In the spec below, what would the resulting PersistentVolumeClaim disk part look like after running through the libvirt defaulting admission controller?
 
```kind: VM
spec:
  domain:
    devices:
      disks:
      - type: PersistentVolumeClaim
        - source:
          name: disk-01
        - target:
          bus: scis
          device: sda
```

rmohr replied:
Libvirt would expand the disk to something like this:

```xml
    <disk type='network' device='disk' snapshot='external'>
      <driver name='qemu' type='raw' cache='none'/>
      <source protocol='iscsi' name='iqn.2017-01.io.kubevirt:sn.42/2'>
        <host name='10.107.27.188' port='3260'/>
      </source>
      <backingStore/>
      <target dev='sda' bus='virtio'/>
      <alias name='virtio-disk0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x04' function='0x0'/>
    </disk>
```

However, the only interesting parts for us are:

```
      <target dev='sda' bus='scis'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x04' function='0x0'/>
```

so the result would look like this:

```yaml
spec:
  domain:
    devices:
      disks:
      - type: PersistentVolumeClaim
        - source:
          name: disk-01
        - target:
          bus: scis
          device: sda
          address:
             type: pci
             domain: 0x00
             bus: 0x00
             slot: 0x04
             function: 0x00
```

We might also want to retrieve the cache, which is `none` by default, however, that also depends on if we treat the absence of `cache` as `none`, or if we want to default it to `none`. But it could also be that we have cases where we want the cache to be influenced by a PVC. In that case we would never default it anywhere. Anyway, the cache value is not important for the Guest OS and therefore does not need to be defaulted by this libvirt service.

davidvossel commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+
+### Migrating between different QEMU versions
+
+In order to make sure that the libvirt version, which does the defaulting
```
Can you clarify this section a bit?

I don't follow how VM spec migration between Qemu versions would take place. If the libvirt defaulting admission controller fills in the spec details before the VM spec is persisted, when would migrating between versions take place.

oourfali replied:
I guess you might need to call this service every time a VM is launched, in order to do the relevant migrations?
Unless we have some way to determine whether an upgrade is needed prior to running the VM?

michalskrivanek replied:
I believe it should be a user initiated action, some sort of a generation number. Similar to cluster level in oVirt.
Once the VM spec is expanded it must not change to the guest and we need to keep it the same. That doesn't necessarily mean everything need to remain, the host-specific part of spec will change from host to host (either live migration or suspend/resume, stop/start flows)

oourfali replied:
Per-VM user operation?
I think the optimal solution would be that this service will be upgraded with other relevant components, and will allow you to "upgrade" the VM spec.

michalskrivanek replied:
we need to allow for up to per-VM granularity, users need to have full control when it happens. Upgrade typically does change the guest environment (e.g. machine types)
It would be reasonable to abstract this to a cluster (~ cluster level compat) or some arbitrary group of VMs in GUI / other administration tools.

oourfali replied:
that probably means adding a version to the VM spec.
but there is still a question who does the upgrade? the same service described here or another.

michalskrivanek commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+a specification which has filled out all fields, which influence how the
+emulated machine looks like for the Guest OS. For instance libvirt network
+configurations may differ on a host to host basis, e.g. because on one host a
+path to an Ethernet device is `/dev/eth0` and on another one `/dev/eth1`,
```
"same network is connected to eth0 on one host and eth1 on another host" ?

michalskrivanek commented on docs/defaulting.md:
````diff
@@ -0,0 +1,249 @@
+```
+
+Finally, when a Domain was successfully started by libvirt, libvirt assigns an
+`alias` to every device in the XML.
````
hm, does it assign aliases on its own? It would be a good idea that we define the alias so we can identify the device easily later on

michalskrivanek replied:
or - we can help pushing that bug so we can actually define our aliases. Otherwise the matching is indeed a pain

michalskrivanek commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+migratable domain XMLs, libvirt fills in all defaults when defining a domain,
+which influence how the emulated machine looks like for the Guest OS. We can
+leverage that with a simple REST based service, which defines VMs in an
+isolated libvirt (running libvirt and the small REST service in a Pod). The
```
I would still much rather interact with something like virt-xml than full blown libvirt environment, that's so much more overhead...

michalskrivanek commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+   before asking libvirt for the defaults. Since these fields should not exist
+   on the VM specification at all (see
+   [docs/vm-configuration.md](docs/vm-configuration.md)), not mapping such
+   dummy data back should be easy.
```
any examples what do you have in mind for the 2 iterms above?

rmohr replied:
Ad fields which don't influence the Guest OS view on the VM) For instance the disk driver element is something which does not influence how the device looks like inside the VM:

```
<driver name='qemu' type='raw' cache='none'/>
```

Ad mandatory domain xml fields)  that is a good question. I can't think of a concrete example, since everything which involves the Guest OS view, should be changable by the user. I should probably change it to, "**If** there are mandatory fields ...".

michalskrivanek replied:
ok. but you do want to have these expressed in VM spec...so the comment relates just to the defaulting, right?

fabiand commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+
+Note that a fully populated VM specification in the context of this proposal, is
+a specification which has filled out all fields, which influence how the
+emulated machine looks like for the Guest OS. For instance libvirt network
```
You should probably reference the [`vm-configuration`](https://github.com/kubevirt/kubevirt/blob/master/docs/vm-configuration.md) proposal here which draws the clear line between backend- and frontend configuration of the specification. Thus the ones affecting the guess and the ones affecting the backend.

For the sake of an aligned language it's then easy to say that this defaulting would be limited to the guest facing, thus frontend configuration.

fabiand commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+   provided, or force users to extract additional data from the status field,
+   which they would have to put into the VM specification in subsequent runs,
+   to get exactly the same VM.
+ * Give us more control over what we want to expose in `status`, instead of
```
The `.status` field is independent of this defaulting approach.
I do see your point how defaulting could have an impact on what the `.status` field exposes. But to me that order of reasoning is the wrong way round, because defaulting (especially ahead of time) is a higher-level concept than reflecting a domains status.

We also need to recognize that some flows will not use an ahead-of-time defaulting approach, because it's not needed/suited everywhere.
Thus I'd suggest to keep the `.status` field discussion out of this document.

rmohr replied:
I think it is essential to give us more freedom on how we plan to shape `.status`. Is it really too specific if I say, that it allows us shaping the `.status` with much more freedom?

fabiand replied:
I don't see how the sentence gives us more freedom or how it restricts our freedom :) I'm just saying that to me teh defaulting should be independent of the `.status` field.

Btw the upcoming custom api server work, would allows us to fetch the current status of a domain on-demand, which would prevent us from updating the domain whenever it changes on the host.

fabiand commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+fill out a lot of device details for you. On cluster level this can lead to
+difficulties, since depending on the fact if the VM has ever run before, you
+will have to operate on different specs, since you have to incorporate VM
+specification changes once a VM was started the first time. Furhter, getting
```
`Further`

fabiand replied:
Could you explain which different `different specs` you refer to here?

rmohr replied:
What would be sufficient for you? What I mean, is that you have to either wait asynchronously for additional spec updates from the cluster, which you need to save for later re-runs (either pull the additional data out from a status, or from an updated spec). So you have a VM Spec before the VM is started, one because of the POST, and one because of the first run, the same for PUTs on hotplug. That are the different ones I mean. Is this more understandable?

fabiand replied:
Ah thanks, yes, I understand it better now.

fabiand commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+   since we could match wrong devices.
+ * After VM restarts, different devices or addresses could be assigned to a VM
+   by libvirt
+ * libvirt assigns all that data on the host, either when defining the Domain,
```
Could you explain this problem a little more?

rmohr replied:
What would you like to see here? Is repeating the mentioned points above enough?

fabiand replied:
Oh, I understood it the right way today.

fabiand commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+   behind the back of the user, all computed default values by libvirt, need to
+   be reflected in the VM status.
+
+If your VMs are ephemeral (no migrations, recreating VMs vs. restarting VMs, no
```
Maybe it's good to reverse the logic and move this sentence further up:  The following problems appear if you are working with persistent VMs.

fabiand commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+
+ * The Kubevirt VM spec can contain abstract definitions, like a reference to a
+   volume claim, or a reference to a host network. In such cases most of the
+   time, the device `source` section needs to be replaced by a dummy `source`
```
I do expect that we have more and more abstractions on the VM API level (today for PVC, in future for networks), thus I wonder if libvirt is the right component to do the defaulting on this level.

rmohr replied:
Like with Pods, I would only aim for backend configuration abstractions. The rest can and should IMHO be covered by admission controllers before the defaulter is invoked.

fabiand commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+   inside the defaulter. Then the fully populated `target` section needs to be
+   mapped back (e.g. a PVC does not have to exist, when posting a VM. It can be
+   mocked out, since it only provides the backend configuration).
+ * Fields which don't influence how the Guest OS sees, shall not be defaulted.
```
Please refer to the "backend" definition from the vm-configuration document here.

fabiand commented on docs/defaulting.md:
```diff
@@ -0,0 +1,249 @@
+ * Fields which don't influence how the Guest OS sees, shall not be defaulted.
+   Such fields need to be filtered out.
+ * There exist mandatory Domain XML fields, which we can't allow to be set on
+   the VM specification. In such cases, the defaulter needs to set dummy values
```
Our VM API must not expose fields which are not allowed to be set. But there is work going on to prevent this.

rmohr commented:
@fabiand 

Thx for the detailed feedback.

> But I'm concerned about the how - In previous proposals we clarified that the cluster level VM API must not (directly) contain host specific resources, and thus we decided that the cluster level VM representation can not be equal to the libvirt domain XML, it can not even be a true subset.
This and some hints in the proposal (i.e. to ignore certain fields, or use dummy values, to map back certain abstractions) raises my concerns that using libvirt for defaulting is the right approach.

I really think that "abstractions" on the VM spec, which are front end configurations, should be expanded/filled by admission controllers before they ever reach etcd. Like with Pods. I would prefer if we only delay the resolution for backend configurations (network, disk). There it makes sense to delay it until the host is involved.

I am open for not using libvirt for this. The main reason is, that at the end always libvirt does this, and we need to cope with things filled values by libvirt in every scenario anyway. At the end, if you put the right admission controllers before this admission controller, it will change exactly nothing on a VM.

> I rather wonder if we can not address the problems individually.
I.e. we could work out a way to be able to achieve a 1:1 mapping between VM API and domxml devices to support the hotplug case. It sounds as if the ordering of the devices is more important than their addresses, this might help to find a solution.

We can aim for such simplifications where possible, but in the end this always only works good, when you restrict the user. If you want to allow things like, for device A the user sets the addresses, but not for device B, then the result is not as clear, since libvirt (or our defaulter) would have to take that into account.

> And if we look at the defaulting part of the proposal, then I do not see any way how we could leverage the proposed mechanism to take i.e. values from libosinfo into account to produce OS tuned hardware configurations. IOW I'd rather like to see on the cluster level that we workout a mechanism which has sane defaults, but can also take informations from sources like libosinfo to create specific tuned hardware configurations. And here I think that we need logic like virt-manager has.

Yes, it is not aiming  for that. A libosinfo admission controller would be invoked before this one. Every libosinfo/preset/virt-xml/... layer will need to be constructed in a way, that it produces reasonable domain xmls, which libvirt would always expand/fill/default in the end. No matter if we have it on the cluster too.

> Btw: I'd still assume that the libvirt defaulting will still happen to some degree on the node, transparently to kubevirt.

Yes, and we should aim for that this only happens in a way that we don't have to propagate it to the VM status for most cases. Anyway, values set there should never change what the Guest OS can see (to simplify for users how to use KubeVirt, not because it is not possible), if they are also part of the VM Specification.

rmohr commented:
> And if we look at the defaulting part of the proposal, then I do not see any way how we could leverage the proposed mechanism to take i.e. values from libosinfo into account to produce OS tuned hardware configurations

Just because I have read this confusion now more than once from different people (probably also because of my lack to express everything properly): This libvirt admission controller here, really is just the last admission controller invoked in a row of possible admission controllers. It shall only do in a synchronous way on a post for the frontend VM configuration, what otherwise the host libvirt would inevitably do.

michalskrivanek commented on docs/defaulting.md:
````diff
@@ -0,0 +1,249 @@
+still be added at the host level. the final flow looks like this:
+
+```
+ client       virt-api/apiserver admission-controller 1 libvirt admission controller  libvirt defaulting service
````
it's not just fancy stuff, it's elementary stuff, e.g. virtio-scsi controller in your example above. Libvirt defaults are as bad as QEMU defaults - sadly it's not what any sane product would want to use.

Anyway, I think it sounds ok, I just think we should have started with the admission controllers before the lowest-level libvirt defaulting.

fabiand commented:
> I am open for not using libvirt for this. The main reason is, that at the end always libvirt does this, and we need to cope with things filled values by libvirt in every scenario anyway. At the end, if you put the right admission controllers before this admission controller, it will change exactly nothing on a VM.

"changing nothing on a VM" depends on what we represent in the VM object. I.e. if we do not represent bus addresses, then defaulting performed by libvirt might not even influence the VM API representation.

And it's a question if we want to expose bus addresses. I had the impression that they might not be needed, as long as we can still identify devices.

> This libvirt admission controller here, really is just the last admission controller invoked in a row of possible admission controllers. It shall only do synchronous on a post for the frontend VM configuration, what otherwise the host libvirt would inevitably do.

And after all this talks, I'm now wondering if we need this last defaulting _on the cluster level_.

IIUIC then this proposal - in the end - would be limited to default addresses and models and missing device dependencies.
For the addresses - It sounds (when speaking to @michalskrivanek and @berrange) that they might not be needed to be exposed on the cluster level.
For models and device dependencies - If they are not specified, then we need a VM API / cluster level logic to default them in an _informed_ way, which is tuned towards the VM API.

And if we have this cluster level VM API aware defaulting mechanism, then I do not see a strong reason for this libvirt based approach _on the cluster level_, because it defaults fields which are not represented (addresses) or fields which should be defaulted by another component anyway (models and device deps).

My take away is that we need to think about _what_ we want to expose in the VM API. Then we'll see what we need to extend. I also see that there is agreement that we need defaulting on the cluster level, and that there will be defaultin gon the node level which is not (necessarily) propagated back.
As said earlier - for the hot-plugging case I could imagine that we find a different solution to have a stable device order and adressing, which does not involve to expose the addresses on the VM API.

rmohr commented:
Sure, if nothing which libvirt can default for the front-end configuration, is exposed or needs to be propagated back, it does not make sense to have it.

davidvossel commented:
> For the addresses - It sounds (when speaking to @michalskrivanek and @berrange) that they might not be needed to be exposed on the cluster level.

@fabiand is there anymore information you can provide here about that conversation?

@rmohr  It's possible we've taken too broad of an approach by calling this "libvirt" defaulting. Maybe we should just figure out exactly what the bare minimum amount of information is we need to fill into a VM spec for us to be able to migrate something like Windows. From there we'd build a defaulting admission controller that begins injecting just those parts.

fabiand commented:
> For the addresses - It sounds (when speaking to @michalskrivanek and @berrange) that they might not be needed to be exposed on the cluster level.
> 
> @fabiand is there anymore information you can provide here about that conversation?

The main outcome of this discussions was that the ordering and identity/addressability of devices is important, but that addresses are not necessarily needed for this.
There should be some more talks to really get all the details.

fabiand commented:
Closing this according to the previous comments.

rmohr commented:
You can close it. But that was not the outcome. The outcome was that it might be possible. It is not yet clear how that influences our spec. @michalskrivanek you promised to look into this?

rmohr commented:
If it is not possible we will need something like this, no matter if libvirt does it or not.

michalskrivanek commented:
yep. can we add #needsresearch label perhaps?

fabiand commented:
#370 can now be used to track different pieces of the problem.
